core/lib/multivm/src/versions/vm_latest/old_vm/memory.rs

```diff
@@ -278,19 +278,31 @@ impl<H: HistoryMode> Memory for SimpleMemory<H> {
 
     fn finish_global_frame(
         &mut self,
-        _base_page: MemoryPage,
+        base_page: MemoryPage,
         last_callstack_this: Address,
         returndata_fat_pointer: FatPointer,
         timestamp: Timestamp,
     ) {
-        // Safe to unwrap here, since `finish_global_frame` is never called with empty stack
-        let _current_observable_pages = self.observable_pages.inner().current_frame();
-        let returndata_page = returndata_fat_pointer.memory_page;
-
-        // This is code oracle and some preimage has been decommitted into its memory.
-        // We must keep this memory page forever for future decommits.
-        let is_returndata_page_static =
-            last_callstack_this == CODE_ORACLE_ADDRESS && returndata_fat_pointer.length > 0;
+        let is_returndata_page_static = last_callstack_this == CODE_ORACLE_ADDRESS;
+        let current_observable_pages = self.observable_pages.inner().current_frame();
+
+        // It is possible that the code oracle runs out of gas after decommitting code.
+        // The page with the code needs to be kept but the zero page is returned instead
+        // in that case.
+        let returndata_page = if is_returndata_page_static {
+            heap_page_from_base(base_page).0
```

Comment:
A little bit dangerous - if we modify CodeOracle in the future it may be hard to debug

Reply:
This is true.

I know nothing about CodeOracle. Maybe we could change it instead?

Though at least there is no chance of such a mistake getting into production because then the decommitted code would always be zero and tests would fail.

Reply:
Also, iiuc, if CodeOracle just forwards old memory page with bytecode, it won't be saved in "returndata_page", right?